File: ios/BorgarlandCore/Tests/BorgarlandCoreTests/DecisionsTest.swift
import XCTest
@testable import BorgarlandCore

/// #89. Both of these decide something a person meets, and neither had a test:
/// they lived in `ReportModel`, which has no test target, so a revert of #73 or
/// #76 left every test in the repository green.
final class DecisionsTest: XCTestCase {

    // MARK: - What the relay's answer means

    func testTheRelaysOwnJudgementIsTrusted() {
        XCTAssertEqual(RelayDisposition.of(status: 201, ok: true), .sent)
        XCTAssertEqual(RelayDisposition.of(status: 200, ok: true), .sent)
    }

    /// The half that costs a report if it is wrong in one direction and a work
    /// queue if it is wrong in the other.
    func testAnAnswerThatWouldBeTheSameNextTimeStopsTheReportWaiting() {
        XCTAssertEqual(RelayDisposition.of(status: 400, ok: false), .refused)
        XCTAssertEqual(RelayDisposition.of(status: 413, ok: false), .refused)
        XCTAssertEqual(RelayDisposition.of(status: 409, ok: false), .refused)
    }

    func testNoAnswerAndTryLaterBothKeepTheReport() {
        // 0 is the transports' convention for nothing answering at all.
        XCTAssertEqual(RelayDisposition.of(status: 0, ok: false), .waiting)
        XCTAssertEqual(RelayDisposition.of(status: 408, ok: false), .waiting)
        XCTAssertEqual(RelayDisposition.of(status: 429, ok: false), .waiting)
        XCTAssertEqual(RelayDisposition.of(status: 500, ok: false), .waiting)
        XCTAssertEqual(RelayDisposition.of(status: 503, ok: false), .waiting)
    }

    func testAnUnrecognisedStatusKeepsTheReportRatherThanDroppingIt() {
        // Failing towards keeping it: a report kept is a report somebody can
        // still send, and a report dropped is gone.
        XCTAssertEqual(RelayDisposition.of(status: 302, ok: false), .waiting)
        XCTAssertEqual(RelayDisposition.of(status: 999, ok: false), .waiting)
    }

    // MARK: - Where the permission stands

    func testGrantedIsGrantedWhateverElseIsTrue() {
        XCTAssertEqual(LocationPermission.of(granted: true, canAskAgain: false), .granted)
        XCTAssertEqual(LocationPermission.of(granted: true, canAskAgain: true), .granted)
    }

    /// The distinction #76 exists for. Both are "not granted"; only one can be
    /// fixed by asking again.
    func testARefusalTheSystemWillRevisitIsNotTheSameAsOneItWillNot() {
        XCTAssertEqual(LocationPermission.of(granted: false, canAskAgain: true), .unanswered)
        XCTAssertEqual(LocationPermission.of(granted: false, canAskAgain: false), .deniedForGood)
    }

    func testTheScreenIsNeverOfferedARetryThatCannotWork() {
        XCTAssertEqual(LocationPermission.granted.exit, .carryOn)
        XCTAssertEqual(LocationPermission.unanswered.exit, .askAgain)
        XCTAssertEqual(LocationPermission.deniedForGood.exit, .openSystemSettings)
    }

    func testTheWalkResumesOnlyWhenTheReasonItStoppedIsGone() {
        XCTAssertTrue(LocationPermission.shouldResume(after: .deniedForGood, nowGranted: true))
        XCTAssertFalse(LocationPermission.shouldResume(after: .deniedForGood, nowGranted: false))
        // Coming back from somewhere else must not restart the location step
        // under somebody who has moved on.
        XCTAssertFalse(LocationPermission.shouldResume(after: .unanswered, nowGranted: true))
        XCTAssertFalse(LocationPermission.shouldResume(after: .granted, nowGranted: true))
    }
}

File: android/app/src/test/kotlin/is/borgarland/DecisionsTest.kt
package `is`.borgarland

import `is`.borgarland.data.LocationPermission
import `is`.borgarland.data.RelayDisposition
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #89. Both of these decide something a person meets, and neither had a test:
 * they lived in PocViewModel, which needs an Application and therefore has no
 * plain JVM test, so a revert of #76 left every test in the repository green.
 * The Swift counterpart is DecisionsTest.swift, case for case.
 */
class DecisionsTest {

    @Test
    fun `the relays own judgement is trusted`() {
        assertEquals(RelayDisposition.SENT, RelayDisposition.of(201, true))
        assertEquals(RelayDisposition.SENT, RelayDisposition.of(200, true))
    }

    @Test
    fun `an answer that would be the same next time stops the report waiting`() {
        assertEquals(RelayDisposition.REFUSED, RelayDisposition.of(400, false))
        assertEquals(RelayDisposition.REFUSED, RelayDisposition.of(413, false))
        assertEquals(RelayDisposition.REFUSED, RelayDisposition.of(409, false))
    }

    @Test
    fun `no answer and try later both keep the report`() {
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(0, false))
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(408, false))
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(429, false))
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(500, false))
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(503, false))
    }

    @Test
    fun `an unrecognised status keeps the report rather than dropping it`() {
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(302, false))
        assertEquals(RelayDisposition.WAITING, RelayDisposition.of(999, false))
    }

    @Test
    fun `granted is granted whatever else is true`() {
        assertEquals(LocationPermission.GRANTED, LocationPermission.of(granted = true, canAskAgain = false))
        assertEquals(LocationPermission.GRANTED, LocationPermission.of(granted = true, canAskAgain = true))
    }

    /**
     * The distinction #76 exists for. Both are "not granted"; only one can be
     * fixed by asking again.
     */
    @Test
    fun `a refusal the system will revisit is not the same as one it will not`() {
        assertEquals(LocationPermission.UNANSWERED, LocationPermission.of(granted = false, canAskAgain = true))
        assertEquals(LocationPermission.DENIED_FOR_GOOD, LocationPermission.of(granted = false, canAskAgain = false))
    }

    @Test
    fun `the screen is never offered a retry that cannot work`() {
        assertEquals(LocationPermission.Exit.CARRY_ON, LocationPermission.GRANTED.exit)
        assertEquals(LocationPermission.Exit.ASK_AGAIN, LocationPermission.UNANSWERED.exit)
        assertEquals(LocationPermission.Exit.OPEN_SYSTEM_SETTINGS, LocationPermission.DENIED_FOR_GOOD.exit)
    }

    @Test
    fun `the walk resumes only when the reason it stopped is gone`() {
        assertTrue(LocationPermission.shouldResume(LocationPermission.DENIED_FOR_GOOD, nowGranted = true))
        assertFalse(LocationPermission.shouldResume(LocationPermission.DENIED_FOR_GOOD, nowGranted = false))
        assertFalse(LocationPermission.shouldResume(LocationPermission.UNANSWERED, nowGranted = true))
        assertFalse(LocationPermission.shouldResume(LocationPermission.GRANTED, nowGranted = true))
    }
}
